### PR TITLE
Inconsistent pillar data from git_pillar (2019.2.x edition)

### PR DESCRIPTION
**Pull request by sathieu**
Same as #57540, ported to Salt 2019.2.x.

### CODE REVIEW

**Comment by waynew** [comment]
No need to port backwards anymore - just the PR against the main/master branch is fine :+1:

**Comment by sathieu** [comment]
I know, but this backport was need by me and others.

**Comment by poofyteddy** [comment]
> No need to port backwards anymore - just the PR against the main/master branch is fine +1

This is the exact reason why we, as a company, are leaving salt.
There is no LTS ... all bug i have found, sometime fixed, never where backported to supported version. sometime they where never ported the the actual version and was waiting for another major update.

We are not talking feature request or breaking change, but to do what the doc say we can do in the version we are running :(

@sathieu thank you for this backport, we are upgrading to 2019 right now and i will give give you feedback about it ;)

File: salt/pillar/git_pillar.py [hunk 1]
```diff
@@ -365,12 +365,9 @@
 from salt.exceptions import FileserverConfigError
 from salt.pillar import Pillar
 
-# Import third party libs
-from salt.ext import six
-
-PER_REMOTE_OVERRIDES = ('env', 'root', 'ssl_verify', 'refspecs')
-PER_REMOTE_ONLY = ('name', 'mountpoint', 'all_saltenvs')
-GLOBAL_ONLY = ('base', 'branch')
+PER_REMOTE_OVERRIDES = ('base', 'env', 'root', 'ssl_verify', 'refspecs', 'fallback', 'saltenv_whitelist', 'saltenv_blacklist', 'disable_saltenv_mapping', 'ref_types,)
```
Comment (poofyteddy) [inline salt/pillar/git_pillar.py:368]: This line end by 'ref_types,) which cause a syntax error

Comment (sathieu) [reply salt/pillar/git_pillar.py:368]: Yes a single quote is missing: `'ref_types,)`.

Question: are you using python2 or python3?

Comment (poofyteddy) [inline salt/pillar/git_pillar.py:368]: `'fallback'` isn't yet available in 2019, this cause an error loading git_pillar

> 2020-06-25 11:05:45,429 [salt.utils.gitfs :2226][CRITICAL][13627] Key 'git_pillar_fallback' not present in global configuration. This is a bug, please report it.
> 2020-06-25 11:05:45,430 [salt.master      :642 ][CRITICAL][13627] Failed to load git_pillar
> 2020-06-25 11:05:45,430 [salt.master      :643 ][CRITICAL][13627] Master failed pre flight checks, exiting

solved when removed

Comment (poofyteddy) [reply salt/pillar/git_pillar.py:368]: python 2, we still have some old server without python3 to manage, this is why i don't upgrade right up to 3001

File: salt/utils/gitfs.py [hunk 2]
```diff
@@ -3150,6 +3208,94 @@ def link_mountpoint(self, repo):
...
+    def __new__(cls, opts, remotes=None, per_remote_overrides=(),
+                per_remote_only=PER_REMOTE_ONLY, global_only=GLOBAL_ONLY, git_providers=None,
+                cache_root=None, init_remotes=True):
+        return GitBase.__new__(
+            self,
+            opts,
+            remotes,
+            per_remote_overrides=per_remote_overrides,
+            per_remote_only=per_remote_only,
+            global_only=global_only,
```
Comment (poofyteddy) [inline salt/utils/gitfs.py:3227]:
```
2020-06-25 11:55:05,166 [salt.pillar      :997 ][ERROR   ][13864] Exception caught loading ext_pillar 'git':
  File "/usr/lib/python2.7/dist-packages/salt/pillar/__init__.py", line 987, in ext_pillar
    key)
  File "/usr/lib/python2.7/dist-packages/salt/pillar/__init__.py", line 915, in _external_pillar_data
    *val)
  File "/usr/lib/python2.7/dist-packages/salt/pillar/git_pillar.py", line 410, in ext_pillar
    global_only=GLOBAL_ONLY)
  File "/usr/lib/python2.7/dist-packages/salt/utils/gitfs.py", line 3222, in __new__
    self,

2020-06-25 11:55:05,166 [salt.pillar      :1049][CRITICAL][13864] Pillar render error: Failed to load ext_pillar git: global name 'self' is not defined

```

Comment (sathieu) [reply salt/utils/gitfs.py:3227]: Replace `self` with `cls` line 3222.

**Comment by sathieu** [comment]
@poofyteddy I've updated branch sathieu:git_pillar_fast_2019 to ec624e7000, with your comments.

**Comment by Vaarlion** [comment]
Hi, i am also trying this fix over the one from @gmsoft-tuxicoman 
Sadly i am getting a bug that i couldn't explain by looking at the diff.
what follow have been anonymised for obvious reason :)

I have a pillar file that contain a line like that
`{% import_yaml 'prod/site/entity/sites_spec.yaml' as yaml_sites_spec %}`
This work well with the original PKG, but return a "file not found" error after patching with the latest version of git_pillar_fast_2019 

```
2020-06-26 15:25:59,649 [salt.pillar      :1049][CRITICAL][18305] Pillar render error: Rendering SLS 'prod.site.entity.server1' failed. Please see master log for details.
2020-06-26 15:26:29,588 [salt.utils.templates:180 ][ERROR   ][18319] Rendering exception occurred
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 169, in render_tmpl
    output = render_str(tmplstr, context, tmplpath)
  File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 443, in render_jinja_tmpl
    trace=tracestr)
SaltRenderError: Jinja error: prod/site/entity/sites_spec.yaml
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 394, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/local/lib/python2.7/dist-packages/jinja2/environment.py", line 1008, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/local/lib/python2.7/dist-packages/jinja2/environment.py", line 780, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 1, in top-level template code
  File "/usr/lib/python2.7/dist-packages/salt/utils/jinja.py", line 193, in get_source
    raise TemplateNotFound(template)
TemplateNotFound: prod/site/entity/sites_spec.yaml
```
After looking at what happend with salt master in debug mod, i found that the file where missing from salt cache when the patch was applied.

```
Vaaelion@salt:~$ ls /var/cache/salt/master/git_pillar/refs/ce98a61442b3db83f0fe3ddb7bd45a7c8e673003abd7c238748adeca2827b2f9/base/prod/site/entity #With the patch
common_prod_mysql.sls  common_prod.sls  common_prod_web.sls  srv-mysql1.sls  srv-mysql2.sls  srv-publish.sls  srv-redis1.sls  srv-redis2.sls  srv-web1.sls  srv-web2.sls  srv-webpp1.sls


Vaaelion@salt:~$ ls /var/cache/salt/master/git_pillar/refs/ce98a61442b3db83f0fe3ddb7bd45a7c8e673003abd7c238748adeca2827b2f9/base/prod/site/entity #Without the patch
common_prod_mysql.sls  common_prod.sls  common_prod_web.sls sites_spec.yaml srv-mysql1.sls  srv-mysql2.sls  srv-publish.sls  srv-redis1.sls  srv-redis2.sls  srv-web1.sls  srv-web2.sls  srv-webpp1.sls
```
I was diging a bit but i didn't find anything that could filer the type/extention of the file in the diff of this patch ...
I hope you'll find the root cause, i sadly need this to be up and running and can't test much. I will apply #54097 once more for now

**Pull request closed without merging**